pkg/master/controller/crdregistration/crdregistration_controller.go

```diff
@@ -195,26 +198,26 @@ func (c *crdRegistrationController) handleVersionUpdate(groupVersion schema.Grou
 		return err
 	}
 	for _, crd := range crds {
-		if crd.Spec.Version == groupVersion.Version && crd.Spec.Group == groupVersion.Group {
-			found = true
-			break
+		if crd.Spec.Group != groupVersion.Group {
+			continue
+		}
+		for index, version := range crd.Spec.Versions {
+			if version.Name != groupVersion.Version || !version.Served {
+				continue
+			}
+			c.apiServiceRegistration.AddAPIServiceToSync(&apiregistration.APIService{
+				ObjectMeta: metav1.ObjectMeta{Name: apiServiceName},
+				Spec: apiregistration.APIServiceSpec{
+					Group:                groupVersion.Group,
+					Version:              groupVersion.Version,
+					GroupPriorityMinimum: 1000,                                        // CRDs should have relatively low priority
+					VersionPriority:      int32(100 + len(crd.Spec.Versions) - index), // CRDs should have relatively low priority
```

Comment:
we want this to result in 100 for CRDs with exactly one version to avoid dueling old API servers

Reply:
I replied there for another suggestion. What if we add `Priority` field to `CustomResourceDefinitionVersion` type? If the user does not provide `Priority` then we can fill it for them (if that is an accepted pattern in our APIs). By having `Priority` field, there is a way for our users to keep order in any way they want.

Reply:
> What if we add Priority field to CustomResourceDefinitionVersion type?

whether it is computed or explicitly set, there is still the possibility of the two versions having flipped priorities in two CRDs. we need to make the outcome of that situation deterministic.

Reply:
Let's step back a bit here:

**Why can different CRD resources have different version orders?**

I want to argue that we should not support this. The same way that names must not overlap between different CRD objects, neither should versions have a mismatch (= being in different order; of course it is fine that different CRDs have different version subsets, but the order should match). 

Any algorithm to create a stable order is wrong for the CRD author in certain cases: the order will potentially not match his intent.

With names we postponed the verification of the not-overlapping invariant into the NamingController which sets the `NamesAccepted` condition. Clearly, it can only use a temporary snapshots of the CRDs it knows (fromt the lister), so we don't have consistency there as we don't have transactions.

I want to argue further that version order is far less critical than names. Hence:

**Why don't we add a `VersionsAccepted` condition from a VersionsController and enforce matching version orders?**

Reply:
> I'm fine with the controller as long as it's HA safe...

It is as safe as the existing naming controller (i.e. there is a race in HA). The assumption is that CRDs are statically created by developers. Even if in an HA environment it can go wrong (= a conflict is not detected and the new CRD versions are served) in 1 out of 100 cases, it will be noticed in 99 cases, and the developer will fix the order.

On the other hand, an algorithm will silently compute some order (which does not match the intend) and the developer of the API group (assuming it is one owner of the group) will not notice. I am not sure I want to trade HA consistency with having a behaviour that the user expects and understands.

Reply:
Stefan's makes a good point about a stable order algorithm leading to action at a distance.  If you modify your CRD with another version and create a conflict, the stable ordering algorithm may make what you're looking at succeed while breaking something else.

By having a proper condition for it, the action happens for *your* change.  You don't see what you expect, you check status and you see why.

It'll be fairly hard to take back the choice and fix behavior later (see TPR to CRD).  I think we should learn from the naming problems and take the proper status an reconciliation approach.